### DISCUSSION TITLE
Reader deprecation

### DISCUSSION THREAD

**Discussion by davidmartos96** [rrousselGit/riverpod#1639] [Q&A]
Hello,
I've just seen the deprecation of Reader in favor of Ref.

I would like to know what would be the preferred way to migrate unit tests that rely on the Reader interface from ProviderContainer.
Example:

```dart
test('some test', () async {
    final pc = ProviderContainer(overrides: [...]);
    final input = "....";
    final result = await functionToTest(pc.read, input);
    ...
}):
```

How would I get a `Ref` in this situation in order to pass it to `functionToTest` as a parameter?

**Comment by rrousselGit** [comment]
You probably should pass the content of the providers instead of container.read

**Comment by TimWhiting** [comment]
Alternatively you can create your own `Reader` type using a typedef. Just look at the type of the `pc.read` function that you are tearing off.

**Reply by rrousselGit** [reply]
Although I wouldn't recommend it.

**Reply by davidmartos96** [reply]
Yes, I could do that. But I thought it would be better to ask first. Is there any specific reason why it's being deprecated? The typedef is still used internally by riverpod 2.0 is it not?

**Reply by rrousselGit** [reply]
It's not used by Riverpod, no.

It's generally a bad idea to build code that abstracts the ref.  
The reasoning is similar to that of not having a shared interface between Ref/WidgetRef. 

The Ref typedef also makes static analysis much harder, which matters for tools like riverpod_graph

**Reply by fujidaiti** [reply]
Is [this](https://riverpod.dev/docs/concepts/combining_providers/#can-i-read-a-provider-without-listening-to-it) code in the tutorial also deprecated? So, should I pass `ref` instead of `ref.read`, or use `ref.read` in provider?

**Reply by rrousselGit** [reply]
Yes. I didn't notice that this article was out of date. I'll update it
